rfc-drafts/stream.md

```diff
@@ -235,27 +281,137 @@ Designing such a migration feature is out of scope for this RFC.
 
 ### IntoStream
 
+**Iterators**
+
 Iterators have an `IntoIterator` that is used with `for` loops to convert items of other types to an iterator.
 
+```rust
+pub trait IntoIterator where
+    <Self::IntoIter as Iterator>::Item == Self::Item, 
+{
+    type Item;
+
+    type IntoIter: Iterator;
+
+    fn into_iter(self) -> Self::IntoIter;
+}
+```
+
+Examples taken from the Rust docs on [for loops and into_iter]](https://doc.rust-lang.org/std/iter/index.html#for-loops-and-intoiterator)
+
 * `for x in iter` uses `impl IntoIterator for T`
+
+```rust
+let values = vec![1, 2, 3, 4, 5];
+
+for x in values {
+    println!("{}", x);
+}
+```
+
+Desugars to:
+
+```rust
+let values = vec![1, 2, 3, 4, 5];
+{
+    let result = match IntoIterator::into_iter(values) {
+        mut iter => loop {
+            let next;
+            match iter.next() {
+                Some(val) => next = val,
+                None => break,
+            };
+            let x = next;
+            let () = { println!("{}", x); };
+        },
+    };
+    result
+}
+```
 * `for x in &iter` uses `impl IntoIterator for &T`
 * `for x in &mut iter` uses `impl IntoIterator for &mut T`
 
+**Streams**
+
 We may want a trait similar to this for `Stream`. The `IntoStream` trait would provide a way to convert something into a `Stream`.
 
 This trait could look like this:
 
-[TO BE ADDED]
+```rust
+pub trait IntoStream where 
+    <Self::IntoStream as Stream>::Item == Self::Item,
+{
+    type Item;
+
+    type IntoStream: Stream;
+
+    fn into_stream(self) -> Self::IntoStream;
+}
+```
 
 ### FromStream
 
+**Iterators**
+
 Iterators have an `FromIterator` that is used to convert iterators into another type.
 
+```rust
+pub trait FromIterator<A> {
+
+    fn from_iter<T>(iter: T) -> Self
+    where
+        T: IntoIterator<Item = A>;
+}
+```
+
+It should be noted that this trait is rarely used directly, instead used through Iterator's collect method ([source](https://doc.rust-lang.org/std/iter/trait.FromIterator.html)).
+
+```rust
+pub trait Interator {
+    fn collect<B>(self) -> B
+    where
+        B: FromIterator<Self::Item>,
+    { ... }
+}
+```
+
+Examples taken from the Rust docs on [iter and collect]](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect)
+
+
+```rust
+let a = [1, 2, 3];
+
+let doubled: Vec<i32> = a.iter()
+                         .map(|&x| x * 2)
+                         .collect();
+
+```
+
+**Streams**
+
 We may want a trait similar to this for `Stream`. The `FromStream` trait would provide way to convert a `Stream` into another type.
 
 This trait could look like this:
 
-[TO BE ADDED]
+```rust
+pub trait FromStream<A> {
+
+    fn from_stream<T>(iter: T) -> Self
+    where
+        T: IntoStream<Item = A>;
+}
```

Comment:
In [`async_std::stream::FromStream`](https://docs.rs/async-std/1.6.2/async_std/stream/trait.FromStream.html#tymethod.from_stream) the `from_stream` method emulates an `async fn`. This differs from `IntoStream` because:

- Going from `T -> Stream` can be sync because the stream itself is async.
- Going from `Stream -> T` (this trait) the yielded value *isn't* async, which means the method to create it needs to be.
```suggestion
pub trait FromStream<A> {
    async fn from_stream<T>(stream: T) -> Self
    where
        T: IntoStream<Item = A>;
}
```

Reply:
It's probably also worth noting that this family of traits may require a fair share of `Send` and `Sized` bounds ([ref](https://github.com/async-rs/async-std/pull/665)), but that's probably out of scope for this section.